### DISCUSSION TITLE
[Done] My API Feedback

### DISCUSSION THREAD

**Discussion by lokka30** [ArcanePlugins/Treasury#117] [Ideas]
I have been writing a test economy provider plugin so I have been able to see what it's like to use Treasury. So far, these are my thoughts on what might need changing.

### To-do list
- [x] Currency#isDefault should probably be named Currency#isPrimary, to suit the other methods in the EconomyProvider interface which are named 'primary'. 
- [x] EconomyFailureReason seems to miss out on a few things. I'll brainstorm these.
- [x] Currency#getDisplayName should probably be named Currency#getDisplayNameSingular, to pertain to Currency#getDisplayNamePlural.
- [x] I'm unsure about Account Permissions being TriStates (true/false/unspecified), I think a simple dual state (true/false) might be better. How is a plugin meant to react when the player isn't known to have or not have a permission? I think it should just default to false in this case.
- [x] I think transactions should have a method that states if they are a 'common' transaction or not. E.g. a common transaction could be the JobsReborn plugin rewarding a user for a task. Then an economy provider plugin can choose not to store these common transactions in a database to save storage space.

### Scrapped ideas
- **Scrapped:** ~~I think Currency#to should be a default method, using MrNemo's formula from a month or two ago. This will also require adding a Currency#getRelativeValue method, e.g. dollars are worth 1.0 and tokens are worth 100.0, so tokens are 100x more valuable than dollars. needs to work with decimal amounts too like 0.5, that's where MrNemo's formula comes in.~~
- **Scrapped:** ~~Ivan, didn't you create a currency parsing method? Because Currency#parse is not a `default` method. It's no biggie if we decide to remove it but it's unfortunate that all that time spent creating it, and it's no longer here?~~

**Comment by lokka30** [comment]
Tagging users who may be interested:
@MrNemo64 @MrIvanPlays @Jikoo

**Reply by MrNemo64** [reply]
> Currency#isDefault should probably be named Currency#isPrimary, to suit the other methods in the EconomyProvider interface which are named 'primary'.

Agree here

> I'm unsure about Account Permissions being TriStates (true/false/unspecified), I think a simple dual state (true/false) might be better. How is a plugin meant to react when the player isn't known to have or not have a permission? I think it should just default to false in this case.

I suppose that asking if a player has a permission and the response is `unknown` it kinda defeats the existence of permissions. You either have them or not

> I think Currency#to should be a default method, using MrNemo's formula from a month or two ago. This will also require adding a Currency#getRelativeValue method, e.g. dollars are worth 1.0 and tokens are worth 100.0, so tokens are 100x more valuable than dollars. needs to work with decimal amounts too like 0.5, that's where MrNemo's formula comes in.

Since anyone can register its own currency a common method to convert between currencies makes sense. If two plugins register its own currency the conversion between them isn't specified

> Ivan, didn't you create a currency parsing method? Because Currency#parse is not a default method. It's no biggie if we decide to remove it but it's unfortunate that all that time spent creating it, and it's no longer here?

In my opinion, "You spent so much time time making it that not having it is a pitty" leads to bad features being added. Although in this case I'd like to see the method `CurrencyAndAmmount parseCurrencyAndAmmount(String str)` come back. Maybe no that names but I'd like it to be back

Nothing to add to the other points

**Reply by lokka30** [reply]
> I suppose that asking if a player has a permission and the response is `unknown` it kinda defeats the existence of permissions. You either have them or not

Exactly

> Since anyone can register its own currency a common method to convert between currencies makes sense. If two plugins register its own currency the conversion between them isn't specified

I think we had a system like this before, no? Perhaps it was lost in the currencies rework.

> In my opinion, "You spent so much time time making it that not having it is a pitty" leads to bad features being added. Although in this case I'd like to see the method `CurrencyAndAmmount parseCurrencyAndAmmount(String str)` come back. Maybe no that names but I'd like it to be back

I see what you mean. I hope that the feature's removal was an accident, and not intentional, since it is very useful.

**Reply by MrNemo64** [reply]
> I think we had a system like this before, no? Perhaps it was lost in the currencies rework.

It was the method that used my formula xd

**Reply by lokka30** [reply]
Yep, you know what I'm talking about :smile:

**Reply by lokka30** [reply]
Update: added to list:

> I think transactions should have a boolean that states if they are a 'common' transaction or not. E.g. a common transaction could be the JobsReborn plugin rewarding a user for a task. Then an economy provider plugin can choose not to store these common transactions in a database to save storage space.

**Reply by MrNemo64** [reply]
> Update: added to list:
> 
> > I think transactions should have a boolean that states if they are a 'common' transaction or not. E.g. a common transaction could be the JobsReborn plugin rewarding a user for a task. Then an economy provider plugin can choose not to store these common transactions in a database to save storage space.

Interesting idea but I think the proposed solution is not ideal. We can assign an importance to a Transaction, for example `LOWEST, LOW, NORMAL, HIGH, HIGHEST` or only `LOW, NORMAL, HIGH`. Now a provider can decide what importance a transaction has to have to be saved in the database. A simple boolean doesn't let much room to play with. 
If we end up adding importances, should the methods that get transactions also accept an importance?

**Reply by lokka30** [reply]
I really like the solution you have come up with. :smiley:
I think the importance should be a part of the transaction itself, not paired with it or anything like that.

**Comment by lokka30** [comment]
See [the branch](https://github.com/lokka30/Treasury/tree/solve-issue-119) for the changes so far. (I've ticked them in the original post above)

I will halt until other users comment on the rest.

**Comment by MrIvanPlays** [comment]
Account permissions should be as is. TriState implementation is used by adventure, velocity, sponge, etc. About currencies - i strongly disagreed about the current changes but you opted for them so we're not returning back to the other implementation because now I think it is better as the economy provider can pretty much control anything.

**Reply by lokka30** [reply]
**Moved double post here from @MrIvanPlays:**
> About transactions - thats completely useless.

> Im sorry for double posting but I needed to catch up - I'm on my village and my mobile data ran out and I'm currently using public wifi.

***

Thanks for the feedback Ivan!

> Use TriStates

I'm confused why they (and we) use the TriState, for the same reasons Nemo and I agreed on, but it must be for good reason.
No problems at all and I will revert the commit that switched them out for booleans. 👍🏻 

> Currencies

I'm happy with how currencies are currently done, with the only exceptions being the `to` and `parse` methods. Previously we had MrNemo's `to` method and your `parse` method, I'm not sure why they no longer exist since they were brilliant.

Should we add them back or leave the `to` and `parse` methods all up to the economy provider to handle?

> Transactions suggestion is useless

I believe otherwise - here's my thought, using an example:

* A server has JobsReborn installed.
* A player decides to go on a mining spree.
* Every second they have a little less than a dollar deposited into their account, as they mine ores, stone, etc.

Why is an economy provider expected to handle such 'spammy' transactions with the same importance as other transactions, such as using `/pay` or purchasing goods from the server shop?

Of course the economy provider can do workarounds such as 'only x transactions per 30 seconds are stored', or 'only transactions over $5 are stored'. But this is not ideal and it also requires the plugin to do extra processing.

I believe it's best that a simple importance level is specified with each transaction so an economy provider can easily know which transactions can be safely discarded from the database (transaction history), which can prevent the server from bloating its database.

Are you convinced? :smile:

> Double posting, internet

No worries at all.

**Reply by MrIvanPlays** [reply]
Leave up to the economy provider. And yes I got convinced about the common boolean on transactions.

**Reply by lokka30** [reply]
Thanks for the feedback. All that's left for me to do is revert the TriState->Boolean change, and then I'll submit a PR.

**Reply by lokka30** [reply]
Are you for/against:

> Currency#getDisplayName should probably be named Currency#getDisplayNameSingular, to pertain to Currency#getDisplayNamePlural.

by the way?

**Comment by MrIvanPlays** [comment]
About transactions - thats completely useless. 

Im sorry for double posting but I needed to catch up - I'm on my village and my mobile data ran out and I'm currently using public wifi.

**Comment by lokka30** [comment]
Pushed some new commits. :smiley:

### e05a9e703a629c7d35341637ee6e2d472812fe4c
* switches back to `TriState` for permissions.
* ~~removes (what I think are) some counter-intuitive defaults from `PlayerAccount.java`~~ **I reverted this change in the 3rd commit, somehow I didn't realise that these played an important role.** I apologise for that!

### 4808176aa39e5cb319398a931080d0aa8631f74c
* renames `Currency#getDisplayName` to `Currency#getDisplayNameSingular`, to pertain to `Currency#getDisplayNamePlural`.
* suppresses single deprecation warning in the migrator for using the 'get current API version' method.

### 68378f979c68dced1e0c9485a907b571bb12f7a4
* reverts PlayerAccount default removals from the first commit
* minor corrections with grammar and unused imports.

**Reply by lokka30** [reply]
### Current status of this discussion:
- Hoping to receive feedback from other developers (if possible)
  - [ ] @MrIvanPlays 
  - [ ] @MrNemo64 
  - [ ] @creatorfromhell 
  - [ ] @Jikoo 
- [x] I'll need to brainstorm some additions to the `EconomyFailureReason` enum.

**Reply by lokka30** [reply]
Pushed more commits!

### 0e30d362fd09d98355c3a0c23ba0747fc16dd0cc
* Minor improvements to EconomyFailureReason.

### 44b4ad7749c5da09875ffd2e7e022ab5ae097ebd
*  Add EconomyFailureReason.CURRENCY_ALREADY_REGISTERED
* Improve documentation for EconomyFailureReason
* > Note: Unsure what the `REQUEST_CANCELLED` constant is for, no example methods have been documented with it. Will need to resolve this before release. If it is kept, need to expand on its documentation as the other constants have received. @creatorfromhell I think you added this, do you know what it is/could be used for?

### 50e89a2d1144329eb3f669395c18c5c8029e5110
* Make Account#setBalance's BigDecimal parameter annotated @NotNull. 

### 6c077022c76f5d722c372aa33c61f73b235bccec
* Added missing @NotNull annotations to the Currency interface. 

### cf7ba4eeef4d7223739deb2db06c84fda5557a14
* All `@Nullable` parameters in the `Currency` interface have been converted to now be `@NotNull Optional<...>` parameters instead.

### 8935c9b1f656c38ec6cf0c102294cf986fbcc69b
* Improve description of `Account#setPermissions`.

### 5e8c509bd034d3820b41bd3360019c17c66fb1f6
* Added the `EconomyFailureReason.NUMBER_PARSING_ERROR` constant.
* Done some minor documentation improvements to the `Currency` interface.

**Reply by lokka30** [reply]
Current status has been updated above. My work here is all done assuming nobody suggests any changes / reversions. If ~3 days' time has passed and all the reviews (if any) have been satisfied, then I will merge this.

**Reply by MrNemo64** [reply]
> All @Nullable parameters in the Currency interface have been converted to now be @NotNull Optional<...> parameters instead.

I don't like the idea of having Optional<> on parameters. If everithing that we use optionals with can be null and if it doesn't have it it cant be null, whats the point of having the NotNull and Nullable anotations? 
I support Optionals as return type but not on parameters

**Reply by MrIvanPlays** [reply]
Optionals are best as they enforce that something may be null. I'm all into using optionals.

**Reply by lokka30** [reply]
I think it is imperative that a library plugin enforces developers to not be lazy with nullability. In other plugins, I would agree with you. I don't like how using Optional is a bit messy, but I think that tradeoff is superseded by everyone's IDEs warning developers that they should handle Optionals properly, instead of using `@Nullable` params and the IDE not really caring much about it.

Are you convinced? Or do you still think we should revert to `@Nullable` params? I doubt I will revert it since Ivan also agrees with me but it will be good to gather a vote on the idea.

Thanks :smiley:

**Reply by lokka30** [reply]
Tagging other users who may be interested in the updates made in this branch. Feel free to leave your opinion if you wish. 
To view all changes and commits see [this](https://github.com/lokka30/Treasury/tree/solve-issue-119) branch.

@yannicklamprecht @Geolykt @NoahvdAa

**Reply by yannicklamprecht** [reply]
I'm on @MrNemo64's side `@Nullable` is more user-friendly. I would personally just overload the method with one argument less.
See the external reasons. The SonarQube rule lists all disadvantages you will have when using optional as a parameter.
https://stackoverflow.com/a/31923105/6686790 | https://stackoverflow.com/a/39005452/6686790
https://rules.sonarsource.com/java/RSPEC-3553

**Reply by lokka30** [reply]
Thanks for the feedback. @MrIvanPlays does this change your opinion at all?

**Reply by MrIvanPlays** [reply]
i saw your code and reviewed. your usage of them is wrong so you have to revert them. we can skip discussing whether to use optional or not for now.

**Reply by MrNemo64** [reply]
I'd like to link [this stack overflow](https://stackoverflow.com/questions/26327957/should-java-8-getters-return-optional-type/26328555#26328555) post about optionals as parameters
Also you are basically achieving nothing, you still need to check if the given optional is, itself, null and all the other parameters
Optionals exist to force the developer to handle a null value, not to indicate that something can be null

**Reply by lokka30** [reply]
Thanks guys. I've addressed this in the latest array of commits 👌🏻
